libcxx/include/__ranges/concat_view.h

```diff
@@ -0,0 +1,637 @@
+    return __x.it_ < __y.it_;
+  }
+
+  _LIBCPP_HIDE_FROM_ABI friend constexpr bool operator>(const __iterator& __x, const __iterator& __y)
+    requires(random_access_range<__maybe_const<_Const, _Views>> && ...)
+  {
+    return __x.it_ > __y.it_;
+  }
+
+  _LIBCPP_HIDE_FROM_ABI friend constexpr bool operator<=(const __iterator& __x, const __iterator& __y)
+    requires(random_access_range<__maybe_const<_Const, _Views>> && ...)
+  {
+    return __x.it_ <= __y.it_;
+  }
+
+  _LIBCPP_HIDE_FROM_ABI friend constexpr bool operator>=(const __iterator& __x, const __iterator& __y)
+    requires(random_access_range<__maybe_const<_Const, _Views>> && ...)
+  {
+    return __x.it_ >= __y.it_;
+  }
+
+  _LIBCPP_HIDE_FROM_ABI friend constexpr auto operator<=>(const __iterator& __x, const __iterator& __y)
+    requires((random_access_range<__maybe_const<_Const, _Views>> && ...) &&
+             (three_way_comparable<__maybe_const<_Const, _Views>> && ...))
+  {
+    return __x.it_ <=> __y.it_;
+  }
+
+  _LIBCPP_HIDE_FROM_ABI friend constexpr decltype(auto) iter_move(const __iterator& __it) noexcept(
+
+      ((is_nothrow_invocable_v< decltype(ranges::iter_move), const iterator_t<__maybe_const<_Const, _Views>>& >) &&
+       ...) &&
+      ((is_nothrow_convertible_v< range_rvalue_reference_t<__maybe_const<_Const, _Views>>,
+                                  __concat_rvalue_reference_t<__maybe_const<_Const, _Views>...> >) &&
+       ...))
+
+  {
+    return std::visit(
+        [](const auto& __i) -> __concat_rvalue_reference_t<__maybe_const<_Const, _Views>...> {
+          return ranges::iter_move(__i);
+        },
+        __it.it_);
+  }
+
+  _LIBCPP_HIDE_FROM_ABI friend constexpr void iter_swap(const __iterator& __x, const __iterator& __y)
+
+      noexcept((noexcept(ranges::swap(*__x, *__y))) &&
+               (noexcept(ranges::iter_swap(std::declval<const iterator_t<__maybe_const<_Const, _Views>>>(),
+                                           std::declval<const iterator_t<__maybe_const<_Const, _Views>>>())) &&
+                ...))
+
+    requires swappable_with<iter_reference_t<__iterator>, iter_reference_t<__iterator>> &&
+             (... && indirectly_swappable<iterator_t<__maybe_const<_Const, _Views>>>)
+  {
+    std::visit(
+        [&](const auto& __it1, const auto& __it2) {
+          if constexpr (is_same_v<decltype(__it1), decltype(__it2)>) {
+            ranges::iter_swap(__it1, __it2);
+          } else {
+            ranges::swap(*__x, *__y);
+          }
+        },
+        __x.it_,
+        __y.it_);
+  }
+
+  _LIBCPP_HIDE_FROM_ABI friend constexpr difference_type operator-(const __iterator& __x, const __iterator& __y)
+    requires __concat_is_random_access<_Const, _Views...>
+  {
+    size_t __ix = __x.it_.index();
+    size_t __iy = __y.it_.index();
+
+    if (__ix > __iy) {
+      std::visit(
+          [&](auto& __it_x, auto& __it_y) {
+            __it_x.apply_fn_with_const_index<tuple_size_v<decltype(__x.parent_->__views_)>>(
```

Comment:
GCC complains about missing of `template`. I don't know why Clang correctly parses this, but perhaps we shouldn't rely on this extension at this moment.
```suggestion
            __it_x.template apply_fn_with_const_index<tuple_size_v<decltype(__x.parent_->__views_)>>(
```

Note that `apply_fn_with_const_index` should be renamed.